PE: preparations for a writer outside of Goblin

RaitoBezarius opened the pull request:
Hi there @m4b, as we discussed it in #381. Here's are commits, more in the mind of:

- Preparations for 1.0 and breaking changes
- Enable to write an external writer given enough data

For now, I will focus on making it possible to write final executables, not object files. I believe this may require more efforts from someone who is really interested into writing a PE compiler or PE object files with this ecosystem.

This is a bunch of QoL improvements for the writer that I'm preparing in another repo, if it's good, and I get myself to 0.x something, I will try to see if it makes sense to send it to faerie or something that would be the good place for such a thing.

In the time being, to unblock on the reviews and merging, here's a bunch of easy things.

I make some stuff `pub` on purpose because I don't believe it's a big deal to let the user observe it, please correct me if you disagree, and I can try to find a way to make it better.

As far as I can tell, this PR **has no breaking changes**. I have more in stock which I am fleshing out to make this easy for you.

RaitoBezarius commented on src/pe/data_directories.rs:
```diff
@@ -18,6 +18,18 @@ impl DataDirectory {
     }
+
+    pub fn data<'a>(&self, bytes: &'a [u8], disk_offset: usize) -> error::Result<&'a [u8]> {
+        bytes
```
I should probably do the safe computation once here.

RaitoBezarius replied:
TODO: doc comment.

RaitoBezarius replied:
Done.

RaitoBezarius commented on src/pe/utils.rs:
```diff
@@ -69,7 +69,7 @@ fn section_read_size(section: &section_table::SectionTable, file_alignment: u32)
 }
 
-fn rva2offset(rva: usize, section: &section_table::SectionTable) -> usize {
+pub fn rva2offset(rva: usize, section: &section_table::SectionTable) -> usize {
```
Maybe a doc comment now it's public?

RaitoBezarius replied:
Done.

RaitoBezarius commented on src/pe/certificate_table.rs:
```diff
@@ -95,10 +111,29 @@ pub struct AttributeCertificate<'a> {
+    ) -> Self {
+        Self {
+            // TODO: turn into a fallible conversion.
+            length: (certificate.len() + ATTRIBUTE_CERTIFICATE_HEADER_SIZEOF) as u32,
```
Should be fallible here.

RaitoBezarius replied:
Addressed.

RaitoBezarius commented on src/pe/section_table.rs:
```diff
@@ -55,6 +58,52 @@ fn base64_decode_string_entry(s: &str) -> Result<usize, ()> {
 }
 
 impl SectionTable {
+    pub fn new(
```
I'm not opposed to moving this outside, I suppose.

RaitoBezarius replied:
TODO if kept: doc comment.

RaitoBezarius replied:
Done.

RaitoBezarius commented on src/pe/utils.rs:
```diff
@@ -22,6 +22,23 @@ fn aligned_pointer_to_raw_data(pointer_to_raw_data: usize) -> usize {
+// Performs arbitrary alignment of values
+// based on homogeneous numerical types.
+#[inline]
+pub fn align_to<N>(value: N, align: N) -> N
```
This could be a more global utility, I can send a PR later to clean up manual alignments if needed.

RaitoBezarius replied:
@m4b basically, this is a useful utility for anyone doing alignments, I can duplicate it downstream but it feels a bit weird as anyone doing any PE shit will have to deal with various alignments computations I'd say?

RaitoBezarius replied:
Resolving, let me know if I should do something else.

RaitoBezarius commented:
https://github.com/RaitoBezarius/ifrit will be my model for the meta writer PE, if it graduates, I will send it your way in faerie if it makes sense. :)

m4b commented:
lol i love the name :) so just catching up here again, iiuc, this has no breaking changes, but you'd like this patch merged right?

m4b reviewed (commented):
mostly questions about `pub`

RaitoBezarius commented on src/pe/certificate_table.rs:
```diff
@@ -77,16 +82,27 @@ impl TryFrom<u16> for AttributeCertificateType {
+/// An alternative name for the WIN_CERTIFICATE header structure.
+pub type WindowsCertificateHeader = AttributeCertificateHeader;
+
+pub const ATTRIBUTE_CERTIFICATE_HEADER_SIZEOF: usize =
```
can be rederived via rust sizeof more or less I suppose, but yes, it's good this is pub. This won't change anyway.
will add doc comment.

RaitoBezarius replied:
Done.

RaitoBezarius commented on src/pe/certificate_table.rs:
```diff
@@ -77,16 +82,27 @@ impl TryFrom<u16> for AttributeCertificateType {
 const CERTIFICATE_DATA_OFFSET: u32 = 8;
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 pub struct AttributeCertificate<'a> {
```
ack

RaitoBezarius replied:
Done.

RaitoBezarius commented on src/pe/mod.rs:
```diff
@@ -39,7 +39,7 @@ use log::debug;
 pub struct PE<'a> {
     /// Underlying bytes
-    bytes: &'a [u8],
+    pub bytes: &'a [u8],
```
I have code somewhere that does the following:
```rust
        for section in self
            .pe
            .sections
            .iter()
            .cloned()
            .map(|s| from_section_table(s, &self.pe.bytes).unwrap())
            .chain(ready_sections.into_iter())
```

outside of the PE code.

The helper is defined as:
```rust
pub fn from_section_table<'a>(
    mut table: SectionTable,
    pe_bytes: &'a [u8],
) -> error::Result<Section> {
    let contents = table.data(pe_bytes)?.map(|c| c.into_owned());
    let relocations = table.relocations(pe_bytes)?.collect::<Vec<_>>();
    let table = core::mem::take(&mut table);

    Ok(Section {
        table,
        contents,
        relocations,
    })
}
```

As I think `SectionTable` has no reference to its original PE binary bytes, it makes it hard to go from a table of sections and look into the contents of a section without going via the PE raw bytes.

We can offer a read-only function on the PE itself I suppose if you prefer, but then, `table.data` will have to take the `PE` itself. What do you think?

m4b replied:
let's make a it a method then; it's easier to deprecate a method than a field if this changes somehow.

RaitoBezarius replied:
Done.

RaitoBezarius commented:
Addressed your comments @m4b, please take a look again.

Right now, it's _mostly_ working, except I'm chasing an alignment issue somewhere... which appeared since the split.

m4b commented:
Cool let me know when you found it ! otherwise this is looking pretty good :)

RaitoBezarius commented:
@m4b I believe we are ready to go, I tracked the last bug which was not a bug in Goblin but my tests being overzealous and wanting equality between unaligned certificates and aligned certificates, so I forked out to this sort of things: https://github.com/nix-community/goblin-signing/pull/6/files#diff-85ad85dc9b5387ec57ec5c58ab8880b11f7897b4e4cf3bb78f02907e630891d3R26-R53

JohnScience commented:
> Hi there @m4b, as we discussed it in #381. Here's are commits, more in the mind of:
> 
> * Preparations for 1.0 and breaking changes: this PR should be good to go for a patch release I'd say.
> * Enable to write an external writer given enough data
> 
> For now, I will focus on making it possible to write final executables, not object files. I believe this may require more efforts from someone who is really interested into writing a PE compiler or PE object files with this ecosystem.
> 
> This is a bunch of QoL improvements for the writer that I'm preparing in another repo, if it's good, and I get myself to 0.x something, I will try to see if it makes sense to send it to faerie or something that would be the good place for such a thing.
> 
> In the time being, to unblock on the reviews and merging, here's a bunch of easy things.
> 
> I make some stuff `pub` on purpose because I don't believe it's a big deal to let the user observe it, please correct me if you disagree, and I can try to find a way to make it better.
> 
> As far as I can tell, this PR **has no breaking changes**. I have more in stock which I am fleshing out to make this easy for you.

Since you mention breaking changes, there are many that should be made for `goblin::pe`, in my opinion.

m4b commented on src/pe/data_directories.rs:
```diff
@@ -18,6 +18,25 @@ impl DataDirectory {
+    /// Given a view of the PE binary, represented by `bytes` and the on-disk offset `disk_offset`
+    /// this will return a view of the data directory's contents.
+    /// If the range are out of bands, this will fail with a [`error::Error::Malformed`] error.
+    pub fn data<'a>(&self, bytes: &'a [u8], disk_offset: usize) -> error::Result<&'a [u8]> {
```
I wonder if this should have just returned an `Option` ? sort of like `get()` for arrays; e.g., is the error message actually useful? there's only one parameter passed in, so similar to an index, it means it's the wrong value basically (bad offset). anyway it's already here so it's probably fine, other than it having to allocate for the format.

m4b commented on src/pe/section_table.rs:
```diff
@@ -55,6 +58,60 @@ fn base64_decode_string_entry(s: &str) -> Result<usize, ()> {
+                .last()
+                .map(|last_section| last_section.virtual_address + last_section.virtual_size)
+                .ok_or(0u32)
+                .unwrap();
```
is this unwrap infallible?

m4b reviewed (approved):
I think this is ok, other than the optional thing and the unwrap being infallible (which can be fixed backwards compatibly); I will wait till tomorrow, if no response, will merge; thanks for your patience @RaitoBezarius :)

I will note this does make a lot more functions pub, which is more maintenance burden, but seems good overall so i'm ok with it.

m4b commented:
> > Hi there @m4b, as we discussed it in #381. Here's are commits, more in the mind of:
> > 
> > * Preparations for 1.0 and breaking changes: this PR should be good to go for a patch release I'd say.
> > * Enable to write an external writer given enough data
> > 
> > For now, I will focus on making it possible to write final executables, not object files. I believe this may require more efforts from someone who is really interested into writing a PE compiler or PE object files with this ecosystem.
> > This is a bunch of QoL improvements for the writer that I'm preparing in another repo, if it's good, and I get myself to 0.x something, I will try to see if it makes sense to send it to faerie or something that would be the good place for such a thing.
> > In the time being, to unblock on the reviews and merging, here's a bunch of easy things.
> > I make some stuff `pub` on purpose because I don't believe it's a big deal to let the user observe it, please correct me if you disagree, and I can try to find a way to make it better.
> > As far as I can tell, this PR **has no breaking changes**. I have more in stock which I am fleshing out to make this easy for you.
> 
> Since you mention breaking changes, there are many that should be made for `goblin::pe`, in my opinion. Right now, I'm documenting everything there and pointing out stuff that in my opinion is a bad idea. There are many such things.

if you have some suggestions here feel free to open up a tracker issue/wishlist/things you think are wrong and we can investigate :) especially if you think there are *crucial* things that need fixing and will break backwards compat, now is the time; if they are large overarching changes to the architecture of goblin's PE though, there is realistically not a good chance it will be merged, but hard to say :)

m4b commented:
@RaitoBezarius pinging you just in case you think this is worth getting in :) @kkent030315 might be something you're interested in too :shrug:
